### PR TITLE
Add smoke tests for terra-jupyter-python and terra-jupyter-aou

### PR DESCRIPTION
**Pull request by deflaux**
This PR adds smoke tests for terra-jupyter-python and terra-jupyter-aou to facilitate performing more regular updates to the Python package versions. (A subsequent PR will modify the Dockerfile for terra-jupyter-python and add more test coverage.)

See the results of successful and failed runs of the GitHub action workflow:
* [terra-jupyter-python](https://github.com/deflaux/terra-docker/actions/runs/552405564)
* [terra-jupyter-aou](https://github.com/deflaux/terra-docker/actions/runs/552405565)
* note the output from pytest displayed within the GitHub action log (e.g. `3 passed, 53 warnings`)
* also note that the output from notebook tests, regardless of success or failure, are available for download via the [workflow artifacts](https://docs.github.com/en/actions/managing-workflow-runs/downloading-workflow-artifacts).

### CODE REVIEW

**Comment by deflaux** [comment]
Note that this succeeds in my fork because my fork has the credential configured. In upstream its trying to do an oauth flow because the credential is not yet configured. Please see [the instructions](https://github.com/deflaux/terra-docker/blob/add-smoke-tests/.github/workflows/test-terra-jupyter-python.yml#L5) for configuring the credential so that a BigQuery read of public data is successful.

**Comment by deflaux** [comment]
Updated the workflow to capture results from all notebook-based tests as a workflow artifact, regardless of success or failure of those tests.

You can see the successful run in my fork: https://github.com/deflaux/terra-docker/actions/runs/538130901

File: terra-jupyter-python/smoke_test.py
Comment (rtitle) [inline terra-jupyter-python/smoke_test.py:1]: Is this executed by the GitHub action or just the `ipynb` notebooks?

Comment (deflaux) [reply terra-jupyter-python/smoke_test.py:1]: The GitHub Action https://github.com/deflaux/terra-docker/blob/add-smoke-tests/.github/workflows/test-terra-jupyter-python.yml#L94

File: terra-jupyter-aou/aou_smoke_test.ipynb
Comment (deflaux) [reply terra-jupyter-aou/aou_smoke_test.ipynb:1]: will do!

File: .github/workflows/test-terra-jupyter-aou.yml
Comment (rtitle) [inline .github/workflows/test-terra-jupyter-aou.yml:9]: Nice comments. :) Makes sense, thinking we can use the `broad-dsp-gcr-public` project for this.

File: terra-jupyter-python/smoke_test.py
Comment (rtitle) [reply terra-jupyter-python/smoke_test.py:1]: Ah I see

**Comment by rtitle** [comment]
I think this looks good! Interested to hear @Qi77Qi's take too.

It will be good to have some testing for these images as part of CI in this repo. We do have integration tests in Leonardo for image updates but those are a bit more heavy-weight and decoupled from PRs in this repo.

If this works well for `terra-jupyter-python` and `terra-jupyter-aou` we could follow up with other images like `terra-jupyter-r`, `terra-jupyter-hail`, etc.

File: .github/workflows/test-terra-jupyter-aou.yml
Comment (Qi77Qi) [inline .github/workflows/test-terra-jupyter-aou.yml:26]: can you remove these if you're ready for this to be merged

**Review by Qi77Qi — COMMENTED** [review]
I think one problem is when we update all image versions, building `terra-jupyer-aou` and `terra-jupyter-python` probably will fail becuz base images wouldn't have existed

**Comment by rtitle** [comment]
> I think one problem is when we update all image versions, building terra-jupyer-aou and terra-jupyter-python probably will fail becuz base images wouldn't have existed

That's a good point -- if updating `terra-jupyter-base` for example we tend to bump all versions in the same PR, which would probably cause these github actions to fail.

Wonder if the actions could be smart enough to build the tree of images if there are any changes. 🤔 Alternatively we could just open separate PRs for each dependent image bump (could be annoying).

File: .github/workflows/test-terra-jupyter-aou.yml
Comment (deflaux) [reply .github/workflows/test-terra-jupyter-aou.yml:26]: I'll update the workflow to conditionally build the base image(s), if it doesn't exist. And then for the purpose of the test, there will be a local edit to the Dockerfile 'FROM' clause to refer to the local-only base image.

will do on removing the test branch triggers

File: terra-jupyter-aou/aou_smoke_test.ipynb
Comment (deflaux) [reply terra-jupyter-aou/aou_smoke_test.ipynb]: this is done now

File: .github/workflows/test-terra-jupyter-aou.yml
Comment (deflaux) [reply .github/workflows/test-terra-jupyter-aou.yml]: Okay, this now builds the base images recursively, when missing.

* [workflow result](https://github.com/deflaux/terra-docker/actions/runs/556319095) when base image for terra-jupyter-python is missing
* [workflow result](https://github.com/deflaux/terra-docker/actions/runs/556319094) when base image for terra-jupyter-aou AND base image for terra-jupyter-python are both missing 
* Here’s how I tested missing base images: [test branch diff](https://github.com/deflaux/terra-docker/commit/0b9d98332c15f0601e06a4ae67480f811e6a9b0b)
   * You’ll see in this diff that I also had to comment out JupyterLab extension setup in terra-jupyter-base.
   * That section appears to be [failing](https://github.com/deflaux/terra-docker/runs/1875710096?check_suite_focus=true) in the current base image.

If this looks good to you folks now, I'll go ahead and remove those test branch triggers.

**Comment by deflaux** [comment]
I assume this looks good to you folks and went ahead and removed the workflow triggers that were only needed for test purposes. Its ready to merge - thanks!

**Comment by Qi77Qi** [comment]
is it expected that the 2 workflows are failing for this PR?

**Comment by deflaux** [comment]
> is it expected that the 2 workflows are failing for this PR?

Yes, please see my earlier comment https://github.com/DataBiosphere/terra-docker/pull/199#issuecomment-772115214

**Comment by Qi77Qi** [comment]
@deflaux  it looks like there's only one test using bq, can we comment that out for this PR? and enable it in a follow up PR? one of us can work with you to get the SA set up properly for this repo so that the PR itself won't have broken tests

**Comment by deflaux** [comment]
> @deflaux it looks like there's only one test using bq, can we comment that out for this PR? and enable it in a follow up PR? one of us can work with you to get the SA set up properly for this repo so that the PR itself won't have broken tests

These are the tests that I think would need to be commented out:
* pandas-gbq
* bigquery magic
* tensorflow gfile

Also, when the full tree of images needs to be built, the gcloud credential is needed to be able to successfully pull `marketplace.gcr.io/google/ubuntu1804:latest`, so even if I comment out the above tests, the GitHub Action will appear broken if you try to tag all the images at the same time.

It only takes 5 minutes to set up the credential. I put the full instructions at the top of the workflow and copied them below for your convenience:
1. Create a service account.
2. Give the service account the following permissions within the project: BigQuery User
3. Store its key and project id as GitHub repository secrets GCP_SA_KEY and GCP_PROJECT_ID.

That's it!

**Comment by Qi77Qi** [comment]
I think the question is mainly which project this SA should live in...
If this is not just used for BQ, forget about the previous comment then

File: .github/workflows/test-terra-jupyter-aou.yml
Comment (deflaux) [reply .github/workflows/test-terra-jupyter-aou.yml:9]: > Nice comments. :) Makes sense, thinking we can use the `broad-dsp-gcr-public` project for this.

@Qi77Qi here is the project @rtitle recommended.

**Comment by Qi77Qi** [comment]
I don't have permission to update SA permission in that project.

@dinvlad  we need an SA that allows some github workflow to pull `marketplace.gcr.io/google/ubuntu1804:latest`, and the workflow also tries to access some public BQ data for testing (hence it needs Bigquery User role), any opinion about where this SA should live?

**Comment by dinvlad** [comment]
@Qi77Qi is access to `marketplace.gcr.io/google/ubuntu1804:latest` required only for Trivy? Or more generally to build this image in GitHub Workflows?

Related, which environment do you need to access BQ data from - is that GitHub Workflows also? And is the BQ data private? Is there any sensitive information stored in it?

Thanks!

**Comment by deflaux** [comment]
> @Qi77Qi is access to `marketplace.gcr.io/google/ubuntu1804:latest` required only for Trivy? Or more generally to build this image in GitHub Workflows?
> 
> Related, which environment do you need to access BQ data from - is that GitHub Workflows also? And is the BQ data private? Is there any sensitive information stored in it?
> 
> Thanks!

I have tested this GitHub Action separately in my fork with a service account that I configured. The service account needs no access to anything within the project from which it originates. It only needs role `BigQuery User` to query public data. With just that one role, the service account has sufficient permission to do all the GCP tasks currently included in these tests such as building an image whose base is  `marketplace.gcr.io/google/ubuntu1804:latest`.

**Comment by Qi77Qi** [comment]
@deflaux , GCP_SA_KEY are GCP_PROJECT_ID stored properly now...but workflow is still failing at `>     pandas_gbq.read_gbq(sql, project_id=os.environ['GOOGLE_PROJECT'])`

**Comment by deflaux** [comment]
> @deflaux , GCP_SA_KEY are GCP_PROJECT_ID stored properly now...but workflow is still failing at `> pandas_gbq.read_gbq(sql, project_id=os.environ['GOOGLE_PROJECT'])`

@Qi77Qi  sounds like we are getting closer!

If you look at GitHub Action step `Setup Cloud SDK` it says:
```
Run google-github-actions/setup-gcloud@master
/bin/tar xz --warning=no-unknown-keyword -C /home/runner/work/_temp/6dcdc6f8-9099-4682-8531-4869f73e199c -f /home/runner/work/_temp/36a2c5d8-a4dc-4cf4-aa89-8bd879bb41c9
No credentials provided, skipping authentication
```

Please double check that the value of `GCP_SA_KEY` is the entire contents of the json key file for the service account (not just the value of the private key).

**Comment by gpcarr** [comment]
i'm a bit confused as to how these smoke tests will alert us of package version incompatibilities. i see that the tests are using packages by importing them and running some simple code - will this indicate when a package is out of date? or is the test simply testing that importing the package and running some code works?

**Comment by deflaux** [comment]
> i'm a bit confused as to how these smoke tests will alert us of package version incompatibilities. i see that the tests are using packages by importing them and running some simple code - will this indicate when a package is out of date? or is the test simply testing that importing the package and running some code works?

The later *simply testing that importing the package and running some code works*. The current set of tests cover sanity-checking some packages for which we've encountered incompatibilities in the past.

The idea is to get the test framework in place, and add more to it over time, as needed.

**Review by gpcarr — APPROVED** [review]

**Comment by Qi77Qi** [comment]
thanks @deflaux for the contribution. I'm going to merge this and validate the workflow once merged...will let you know if there's any issue

**Pull request merged**
